src/context.rs

```diff
@@ -0,0 +1,230 @@
+use futures::channel::oneshot;
+use futures::future::{pending, select, BoxFuture, FutureExt, Shared};
+use tokio::sync::mpsc;
+use tokio::time::{delay_for, Duration};
+
+pub struct CancelHandle {
+    cancel_fn: Box<dyn FnOnce()>,
+}
+
+impl CancelHandle {
+    pub fn cancel(self) {
+        let _ = (self.cancel_fn)();
```

Comment:
No need to do the ignoring let binding for a function with a unit return type.

```suggestion
        (self.cancel_fn)();
```